editor/plugins/editor_context_menu_plugin.cpp

```diff
@@ -0,0 +1,210 @@
+			break;
+		}
+	}
+}
+
+void EditorContextMenu::_bind_methods() {
+	BIND_ENUM_CONSTANT(CONTEXT_SLOT_SCENE_TREE);
+	BIND_ENUM_CONSTANT(CONTEXT_SLOT_FILESYSTEM);
+	BIND_ENUM_CONSTANT(CONTEXT_SLOT_SCRIPT_EDITOR);
+}
+
+void EditorContextMenu::handle_plugins(ContextMenuSlot p_slot, const Vector<String> &p_paths, PopupMenu *p_popup) {
+	Ref<EditorContextMenuPlugin> *plugin_list = get_plugin_array(p_slot);
+	bool add_separator = false;
+
+	for (int i = 0; i < MAX_PLUGINS; i++) {
+		Ref<EditorContextMenuPlugin> plugin = plugin_list[i];
+		if (plugin == nullptr) {
+			break;
+		}
+		int submenu = plugin->get_meta("submenu", 0);
+		if (!plugin->can_handle(p_paths) || submenu) {
+			continue;
+		}
+		if (!add_separator) {
+			add_separator = true;
+			p_popup->add_separator();
+		}
+		Array params = plugin->get_meta("parameters");
+		Ref<Texture2D> icon = params[0];
+		int idx = plugin->get_meta("idx");
+		if (!icon.is_null() && icon.is_valid()) {
+			p_popup->add_icon_item(icon, params[1], idx);
+		} else {
+			p_popup->add_item(params[1], idx);
+		}
+	}
+}
+
+//Handle submenu item
+void EditorContextMenu::handle_plugins(ContextMenuSlot p_slot, const Vector<String> &p_paths, PopupMenu *p_popup, EditorContextMenuPlugin::ContextSubMenuSlot p_submenu_slot, PopupMenu *p_submenu) {
+	Ref<EditorContextMenuPlugin> *plugin_list = get_plugin_array(p_slot);
+	bool add_separator = false;
+
+	for (int i = 0; i < MAX_PLUGINS; i++) {
+		Ref<EditorContextMenuPlugin> plugin = plugin_list[i];
+		if (plugin == nullptr) {
+			break;
+		}
+		int submenu = plugin->get_meta("submenu", 0);
+		if (p_submenu == nullptr || submenu != p_submenu_slot) {
+			continue;
+		}
+		if (!plugin->can_handle(p_paths)) {
+			continue;
+		}
+		if (!add_separator) {
+			add_separator = true;
+			p_submenu->add_separator();
+		}
+		Array params = plugin->get_meta("parameters");
+		Ref<Texture2D> icon = params[0];
+		int idx = plugin->get_meta("idx");
+		if (!icon.is_null() && icon.is_valid()) {
+			p_submenu->add_icon_item(icon, params[1], idx);
+		} else {
+			p_submenu->add_item(params[1], idx);
+		}
+	}
+}
+
+template <typename T>
+void EditorContextMenu::invoke_plugin_callback(ContextMenuSlot p_slot, int p_option, const T &p_arg) {
+	Ref<EditorContextMenuPlugin> *plugin_list = get_plugin_array(p_slot);
+	Variant arg = p_arg;
+	Variant *argptr = &arg;
+
+	for (int i = 0; i < MAX_PLUGINS; i++) {
+		Ref<EditorContextMenuPlugin> plugin = plugin_list[i];
+		if (plugin == nullptr) {
+			break;
+		}
+		int idx = plugin->get_meta("idx");
+		if (p_option != idx) {
+			continue;
+		}
+		Array params = plugin->get_meta("parameters");
+
+		Callable callback = params[2];
+		Callable::CallError ce;
+		Variant result;
+		callback.callp((const Variant **)&argptr, 1, result, ce);
+
+		if (ce.error != Callable::CallError::CALL_OK) {
+			String err = Variant::get_callable_error_text(callback, nullptr, 0, ce);
+			ERR_PRINT("Error calling function from context menu: " + err);
+		}
+	}
+}
+//scene_tree_dock
```

Comment:
```suggestion

// scene_tree_dock
```
Space between methods